### ISSUE TITLE
The order of top-level declarations should be irrelevant

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#875]
https://github.com/gpuweb/gpuweb/issues/741 describes how authors usually don't put all of their shader sources in a single file; they often split them up in many source files. That issue seems to be moving in the direction of "authors should just concatenate all their sources together in JS before delivering to the browser."

If order of declarations at the top level is important, this approach breaks down. Traditionally, `#include` order is defined _inside_ source files, so if order is important in WGSL and authors are using their existing workflows, they would either need an authoring-time build step, or a WGSL parser in Javascript, to resolve the include order. Avoiding these things are some of the reasons WGSL exists in the first place.

A better solution is to treat top-level declarations as an unordered set, rather than an ordered array. From an authoring perspective, this means that authors can just concatenate their sources in JS in any order and it will just work (no parser required). From an implementation perspective, resolving names would be pulled out into a separate pass after the parser is completed. This doesn't really increase complexity because it's the same name resolution code - it's just at a different place in the shader pipeline.

If we resolve the order of top-level declarations will be important, such a resolution would indicate that we should take another pass at solving https://github.com/gpuweb/gpuweb/issues/741 a different way.

**Comment by dj2** [comment]
The concatenation approach works fine with #include and #ifdef guards. You have to #include what you're using, and those includes will end up being in order.  This doesn't require a WGSL parser in JS, just a preprocessor.

**Comment by kvark** [comment]
> From an implementation perspective, resolving names would be pulled out into a separate pass after the parser is completed. This doesn't really increase complexity because it's the same name resolution code - it's just at a different place in the shader pipeline.

That also forbids name resolution at parsing time, right? That sounds like a potential increase in complexity for some tools.

**Comment by magcius** [comment]
GLSL also has this problem today, requiring somewhat sophisticated shader assembly tools. To give some examples: `#version` must come at the top for GLSL ES 3.0, followed by `#extension`. If trying to build variants out of a collection of smaller pieces, one commonly runs into errors, e.g. trying to define an `attribute` in an FS shader is disallowed, requiring them to be stripped.

Thankfully, JavaScript has a features to help build a shader assembly pipeline, like regular expressions. I would rather see browser implementation cost be simpler in favor of a more sophisticated author-driven tool pipeline.

**Comment by litherum** [comment]
> The concatenation approach works fine with #include and #ifdef guards.

We don't have #include or #ifdefs (https://github.com/gpuweb/gpuweb/issues/568). Such a solution would require authors to run a preprocessor on their own in JS. Avoiding this requirement is one of the whole reasons WGSL exists in the first place.

**Comment by dj2** [comment]
I don't believe that was the whole reason for WGSL. We've always said if you need preprocessor behaviour do it in JS, @Kangz has shown how a JS preprocessor could work.

**Comment by johnkslang** [comment]
>  #version must come at the top for GLSL ES 3.0

Not that I want to complicate WGSL, just say this was an arbitrary and controversial decision, not followed by desktop GLSL.

WGSL could certainly have a #version-like declaration rules that would allow concatenation of shader pieces.  I would certainly know how to tweak GLSL to allow that for GLSL, for all global declarations.

Just hoping, I guess, to keep reasons for not doing things confined to whatever the real reasons actually are.

**Comment by grorg** [comment]
Discussed at the [2020-06-24 WebGPU Virtual F2F](https://docs.google.com/document/d/1uv0LiQB2gAFbUEU3qx2-Sgsv6r3fEKMvnifbFoRbvmw/).

**Comment by kvark** [comment]
Discussed at the [2021-02-23 WebGPU Virtual F2F](https://docs.google.com/document/d/1JXZ9DhMTq-Y0_132zDX3-46omBKg3wSqgEJLsRZEUkg/edit#heading=h.oeagdcnwhg3d).
The conclusion was to postpone this till after MVP, and land a change to allow locals to shadow globals, in order to be forward-compatible.

**Comment by litherum** [comment]
(... to shadow globals, and possibly locals too...)

**Comment by Jamesernator** [comment]
While trivial concatenation preprocessors are very very easy to implement in JS, it would be nice if WGSL supported "module local" scopes so that utility libraries can reliably use top-level names without concern for collisions.

Obviously the biggest example now would be how ESM in JS has been specified, but even WASM has a similar thing where you can create individual modules each with their own "scope".

Doing this sort've preprocessing isn't that easy in JS-land as you need to construct an entire parser in order to reliably do the variable renaming.

It'd be nice if we had an API similar to `WebAssembly` where we can create uninstantiated shader modules and link them with whatever dependencies we want. This would be a great way of both sharing code amongst many shaders (and the browser can just parse each module once), and a great way to specialize shaders with potentially arbitrary behaviour just by linking the right functions.

e.g. As an example of how it could look:

```
/* mod.wgsl */
[[export]]
fn negate(vec3<f32> vector) -> vec3<f32> {
  return vec3<f32>(-vector.x, -vector.y, -vector.z);
}
```

```
/* other.wgsl */
[[export]]
fn negate(vec2<f32> vector) -> vec2<f32> {
  return vec2<f32>(-vector.x, -vector.y);
}
```

```
/* main.wgsl */
[[import("mod.wgsl", "negate")]]
// Explictly declared allows for AOT optimizations, but this could be elided in favour
// of discovering types at import time
fn negate(vec3<f32> vector) -> vec3<f32>; 

[[import("other.wgsl", "negate")]]
fn negateVec2(vec2<f32> vector) -> vec2<f32>;

[[stage(fragment)]]
fn vertexMain() -> VertexOutput {
  // ...
}
```

```js
// This only needs to be parsed once, and can be included in as many other
// program shaders as wanted
const modShaderModule = webGpu.createShader(mainShaderString);
console.log(modShader.exports); // ["negate"];
const otherShaderModule = webGpu.createShader(otherShaderString);
console.log(otherShader.exports); // ["negate"]

const mainShaderModule = webGpu.createShader(modShaderString);
console.log(mainShader.imports); // ["mod.wgsl", "other.wgsl"];

const mainShader = mainShaderModule.instantiate({
  "mod.wgsl": modShaderModule,
  "other.wgsl": otherShaderModule,
});

// do stuff with mainShader
```

**Comment by kvark** [comment]
@Jamesernator you are essentially suggesting us to have a module system. This is a bigger task than the subject.
And the problem you are trying to solve (avoid naming collisions) seems like a rather small problem. If you hit the collision, you hit it very early - at shader module creation time, and you can work around it by having some naming convention for your "includes".
With this in mind, I'd suggest filing a separate issue and expecting it to be considered after MVP.

**Comment by magcius** [comment]
That also changes the compilation model quite a bit, and seems very out of place for this issue.

**Comment by litherum** [comment]
https://github.com/gpuweb/gpuweb/issues/777 is about namespaces, which can help with name collisions.

**Comment by Jamesernator** [comment]
> you are essentially suggesting us to have a module system. This is a bigger task than the subject. And the problem you are trying to solve (avoid naming collisions) seems like a rather small problem.

> That also changes the compilation model quite a bit, and seems very out of place for this issue.

> #777 is about namespaces, which can help with name collisions.

Namespaces/module systems yes also encapsulate name collisions which is something I'm motivated by, but also the issue was raised because of concatentation behaviour of pseudo-module systems like `#include`. Modules are certainly relevant here as if we had (non-cyclic) modules, the browser could trivially perform concatenation (and module scope renaming, but that's technically separate) in the correct order as it knows the tree-order of linked modules.

And sure this again could be solved in userland, but it kinda begs the question as to why we have a text format if it's simply going to require people to potentially ship multiple tools for "compiling" it together in the first place.

Although I think I will open an issue, because yes modules do cover quite a number of issues, including name collisions, re-usability, no multiple parses, and the issue that was specifically raised here.

**Comment by fintelia** [comment]
My understanding is that in WGSL functions must be defined when they are declared, and global variable bindings must be specified at the declaration site as well. The lack of prototypes combined with order dependence worries me.

The net result seems to be that the "just concatenate the files" strategy is going to feel very limited. You largely can't just prepend a handful of reusable libraries (L1, L2, ...) to a single source file (S) containing the shader entry points:

* S must appear after L1, L2, etc. so that it can use the functions declared in those libraries, however
* Textures and other global variables cannot be defined in library files because the assigned bindings might be incompatible, or worse a single variable might be needed by multiple libraries. Thus S must appear before L1, L2, etc.
* If there are any common struct definitions, those have to appear before the global variables which means they can't be in the libraries.

The best solution I can come up with is to interleave all of these by splitting each library in half so functions and structs can be defined separately, and similarly splitting the main source file so that global variables and the shader entry points can be distinct.

**Comment by litherum** [comment]
I actually like the [module idea](https://github.com/gpuweb/gpuweb/issues/875#issuecomment-828890993), but I have to say that it probably doesn't make the cut for the "minimum viable product." I also do agree that a module system is sufficiently distinct from the title of this issue that it warrants its own issue.

If we wanted to have modules, we would want to make sure they would turn into [`MTLDynamicLibrary`](https://developer.apple.com/documentation/metal/mtldynamiclibrary?language=objc) on Metal backends. `MTLDynamicLibrary` has 2 cool benefits:
1. If two shaders use the same library, that code only exists at runtime in memory once - it doesn't get duplicated in memory for each shader.
2. The code in the `MTLDynamicLibrary` only has to be compiled once, no matter how many `MTLLibrary`s use it, thereby speeding up compilation time.

There are (at least) two significant differences, though, in WGSL:
1. Unlike Metal, WGSL's resources exist at the global scope. If one module can't access globals from another module, that effectively means that modules can't (directly) interact with resources from the entry point. That seems limiting.
2. At least two of the analyses we plan to perform in WGSL - uniformity analysis and filterable texture analysis - must be run on the entire call graph. We would have to figure out whether we have to wait to run these analyses until we have the whole call graph, or whether the analyses can be split up into module-sized chunks.

**Comment by kvark** [comment]
> Textures and other global variables cannot be defined in library files because the assigned bindings might be incompatible, or worse a single variable might be needed by multiple libraries.

I totally do not understand this part. A library is expected to have some control of the bind group layout, so it's free to inject the relevant textures/samplers/etc into the WGSL code, using the controlled bindings. The problem of "single variable used by multiple libraries" is not a problem with WGSL, is it? The libraries can coordinate the way they want to share resources.

**Comment by fintelia** [comment]
Perhaps I should step back and describe how I have been doing shader compilation. My strategy has been to specify a sequence of GLSL source files. One of the source files with contain the entry point, while the others will be libraries that are used by multiple shaders. At runtime, I load each of the files in sequence and concatenate them together to produce a single string. That string is run through a GLSL->SPIRV compiler and then reflected to gather the names and types of each global variable.

Using the results of the reflection, I programmatically produce a single BindGroup and BindGroupLayout. I cannot use the same BindGroup everywhere because some shaders write to textures/buffers that others use read-only (worse, in rare cases a shader will read from one texture array layer while writing to another)

> A library is expected to have some control of the bind group layout, so it's free to inject the relevant textures/samplers/etc into the WGSL code, using the controlled bindings.

With the "just concatenate" strategy, wouldn't control over this would mean the library would directly contain the variable declaration and associated bind group index and location annotations? That would also require that the same ones are used every time that library is included. 

> The problem of "single variable used by multiple libraries" is not a problem with WGSL, is it? The libraries can coordinate the way they want to share resources.

I'm not sure how this coordination can work with simply concatenating files. If I have L1 and L2 that both use a variable, then for a given shader I'd like to have the option include just L1, just L2, or both L1 and L2. The global variable can't be declared in a single of the libraries or else it wouldn't work if only the other was used, but it can't be declared in both because the two declarations would conflict when both files were included.

**Comment by kvark** [comment]
@fintelia thank you for describing the work flow!
> That string is run through a GLSL->SPIRV compiler and then reflected to gather the names and types of each global variable.

How are you seeing this ported to WebGPU? Would you use the SPIR-V result, and convert it to WGSL? Or would you convert the GLSL to WGSL and then try to concatenate them? Note that WebGPU doesn't provide you with a way to reflect the shader/pipeline.

> With the "just concatenate" strategy, wouldn't control over this would mean the library would directly contain the variable declaration and associated bind group index and location annotations? That would also require that the same ones are used every time that library is included.

Correct, yes. The library would know its bind group layout index and all the locations.

> If I have L1 and L2 that both use a variable, then for a given shader I'd like to have the option include just L1, just L2, or both L1 and L2.

If you want to have re-usable modules, then you'd want to move anything shared between modules into another module. I.e. you'd have L3 containing the variable declaration, and then internally consider L1 dependent on L3 and L2 dependent on L3. This way, L3 would only be included once in the contatenated stream.

**Comment by dneto0** [comment]
Question:  How does order independence interact with the 'enable' directive?

> An enable directive indicates that the functionality described by a particular named extension may be used in the source text after the directive itself. That is, language functionality described by the extension may be used in any source text after the enable directive.

The enable directive can introduce new names visible at module-scope.  I expect new types and functions especially.  (I suppose let- and var-declarations too.)   

Example using a function before it should be known:

```rust
 fn bar() -> f32 {   return fooRand(); }

enable foo_utils;  // defines fooRand()
```

Example using a type before it should be known:

```rust
let small = f16(1.0);            // Type-conversion introduced by 'enable f16';
fn smallish() -> f16 { small; }  // Return type introduced by 'enable f16'

enable f16;
```

(It's not material whether 'f16' as a reserved word or not.)

It seems weird/wrong to allow such uses.

Als

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-10-19-Minutes#the-order-of-top-level-declarations-should-be-irrelevant-875">WGSL meeting minutes 2021-10-19</a></summary>



* DN: Nesting of scopes makes this more palatable. However, still tbd: How does `enable` directives work? Do they affect text that came before them
* MM:
    * 1.  ‘Enable’ can be in the middle and only applies after (what we have today)
    * 2. Require `enable` at the top
    * 3. ‘Enable’ can be anywhere but applies globally.
    * 4. `enable` direction to be out of band
* RM: One of goals was to allow stitching of shaders by concatenation, which pushes us towards not requiring enable-at-top and not out-of-band
* RM: If we really want to support stitching of shaders by concatenation, maybe we want a scope for `enable`s
* MM: Maybe use more squiggly braces
* BC: Maybe this makes it tricky, e.g. if `enable` lets you return something new, does the caller need to have it `enable`d also?
* AB: `f16` is a good example, where a new type is introduced. Maybe we could decouple the keyword and identifier namespaces?
* BC: I think at-top is good. It’s unlikely that stitching will work great
* MM: I think that’s reasonable, even if you have a big body of source files, you may still be able to say “these all need X,Y,Z”
* DN: I like the idea of having enabling first. Doing more than that feels like more pain than profit. Worried about having to carry the lexical position all the way through parsing/validation/etc. Could also run into an issue where two extensions e.g. collided on a name
* MM: Question for RM, how do you feel about parsing the source twice.
* RM: Not super happy, probably not a huge deal though. Seems like a big hack, prefer to avoid.
* DM: Can we postpone this feature?
* DM: Changing the parser to 2 pass is a big change for us. (We do name resolution during parsing.)
* …
* DM: Weren’t you going to propose something like module linking to satisfy this MM?
* MM: Why not both? I think we do want both, but also not for mvp.
* MM: Also, using this from Metal requires access to the filesystem today. 
* DM: Maybe this is something that can be fixed before we get to it? :)
* MM: We do have a radar about it
* JP: About postponing, since we already have enable, does it prevent us changing this in the future?
* MM: We might need an `enable` for this :)
* MM: I was thinking of something like array syntax, where if we enable it, we might need to parse *differently*. 
* **BC: Would like to talk to DN about this.**
* MM: Would we like to postpone for a week, or for longer?
* **DN: Next week is fine, though I’d like to see a proposal from MM**


</details>

**Comment by dneto0** [comment]
After discussion offline, our recommendation is:

- require `enable` directives to be stated at the top of the module.  And then trivially they would apply to all the (ordinary) source code.
- each top-level declaration is visible everywhere, i.e. the scope of each declaration is the entire program.
    - require that silly cycles don't exist.  E.g. when we make initializers for let-declarations more flexible, they can't induce a cycle in evaluation.   

Example of a silly cycle (assuming reasonable extension of WGSL):
```
let foo : i32 = bar;
let bar: i32 = foo;   // invalid. cycle.
```
Another silly cycle, assuming there's a consteval-like concept for functions:
```
[[consteval]] fn foo() -> i32 { return bar; }
let bar: i32 = foo();
```

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-11-02-Minutes#the-order-of-top-level-declarations-should-be-irrelevant-875">WGSL meeting minutes 2021-11-02</a></summary>



* DN: We looked offline, some recommendations:
    * 1. Move enable directives to always-at-top
    * 2. When you make a decl, where is it visible? Top-level-decl visible across entire module.
    * Must watch out for “silly cycles”, like declarations that depend on how they’re defined
* RM: Happy with this, and we already have checks like this for recursion, so we don’t see this as an issue.
* JB: Naga also already handles the acyclic check separately from the parser
* **Resolved as above, Needs Spec**


</details>

**Comment by kvark** [comment]
I'm not happy with this decision, since it forces us to effectively rewrite the parser for something that's not available in any of the current shading languages. And this is very late in the shipping process (i.e. we have other more important work to do for shipping).

The original desire for this, as stated, is to support shader concatenation. I believe it's important for the authors to know and understand the dependency between their sources, in which case they can figure out the order of concatenation. Otherwise, do we expect them to slap a bunch of sources together and see if this works as a suggested workflow?
